### PR TITLE
[Raft] Implement heartbeat.

### PR DESCRIPTION
**Pull request by liana-yang**

### CODE REVIEW

File: lab/src/raft/leader_election.go [hunk 1]
```diff
@@ -19,6 +25,40 @@ func (rf *Raft) runLeaderElection() {
...
+		PrevLogTerm:      lastLog.Term,
+		Entries:          []LogEntry{},
+		LeaderCommit:     rf.commitIndex,
+	}
+	reply := AppendEntriesReply{}
+	for i := range rf.peers {
+		if i != rf.me {
+			rf.sendAppendEntries(i, &args, &reply)
+			if reply.Success {
+				// If successful: update nextIndex and matchIndex for follower (§5.3)
```
Comment (flyingmeat) [inline lab/src/raft/leader_election.go]: This should be done inside appendEntries

File: lab/src/raft/leader_election.go [hunk 2]
```diff
@@ -19,6 +25,40 @@ func (rf *Raft) runLeaderElection() {
...
+				rf.matchIndex[i] = lastLog.Index
+				rf.nextIndex[i] = lastLog.Index + 1
+			} else {
+				if reply.Term > rf.currentTerm {
+					// If RPC request or response contains term T > currentTerm:
+					// set currentTerm = T, convert to follower (§5.1)
+					rf.toFollower(reply.Term)
+				} else {
+					// If AppendEntries fails because of log inconsistency: decrement nextIndex and retry (§5.3)
+					rf.nextIndex--
```
Comment (flyingmeat) [inline lab/src/raft/leader_election.go]: ditto

File: lab/src/raft/leader_election.go [hunk 3]
```diff
@@ -19,6 +25,40 @@ func (rf *Raft) runLeaderElection() {
...
+					rf.sendHeartbeat()
```
Comment (flyingmeat) [inline lab/src/raft/leader_election.go]: You dont have to retry here

File: lab/src/raft/append_entries.go [hunk 4]
```diff
@@ -18,20 +18,26 @@ type AppendEntriesReply struct {
 	Success bool
 }
 
+func (rf *Raft) sendAppendEntries(server int, args *AppendEntriesArgs, reply *AppendEntriesReply) bool {
+	ok := rf.peers[server].Call("Raft.AppendEntries", args, reply)
+	return ok
+}
+
 func (rf *Raft) AppendEntries(args *AppendEntriesArgs, reply *AppendEntriesReply) {
 	rf.Lock()
```
Comment (flyingmeat) [inline lab/src/raft/append_entries.go]: typo

File: lab/src/raft/append_entries.go [hunk 5]
```diff
@@ -18,20 +18,26 @@ type AppendEntriesReply struct {
...
-
 	//  Reply false if term < currentTerm (§5.1)
 	if args.Term < rf.currentTerm {
 		reply.Success = false
 		return
 	}
 
+	rf.leader == args.Leader
+	rf.lastHeartBeat = time.Now()
+
```
Comment (flyingmeat) [inline lab/src/raft/append_entries.go]: If the current server is candidate should convert to follower

File: lab/src/raft/leader_election.go (hunk 1 above)
Comment (liana-yang) [reply lab/src/raft/leader_election.go]: Done.

File: lab/src/raft/leader_election.go (hunk 2 above)
Comment (liana-yang) [reply lab/src/raft/leader_election.go]: Done.

File: lab/src/raft/leader_election.go (hunk 3 above)
Comment (liana-yang) [reply lab/src/raft/leader_election.go]: Done.

File: lab/src/raft/append_entries.go (hunk 4 above)
Comment (liana-yang) [reply lab/src/raft/append_entries.go]: Done.

File: lab/src/raft/append_entries.go (hunk 5 above)
Comment (liana-yang) [reply lab/src/raft/append_entries.go]: Done.

File: lab/src/raft/append_entries.go (hunk 4 above)
Comment (flyingmeat) [reply lab/src/raft/append_entries.go]: line 27 is still rf.Lock()

**Review by flyingmeat — APPROVED** [review]

File: lab/src/raft/append_entries.go (hunk 4 above)
Comment (liana-yang) [reply lab/src/raft/append_entries.go]: Done.

**Pull request merged**